package org.apache.hive.io.file.orc;

message Type {
  enum Kind {
    BOOLEAN = 0;
    TINYINT = 1;
    SMALLINT = 2;
    INT = 3;
    BIGINT = 4;
    FLOAT = 5;
    DOUBLE = 6;
    STRING = 7;
    BINARY = 8;
    DATETIME = 9;
    ARRAY = 50;
    MAP = 51;
    STRUCT = 52;
    UNION = 53;
  }
  required Kind kind = 1;
  repeated uint32 subtypes = 2 [packed=true];
  repeated string fieldNames = 3;
}

message StripeInformation {
  optional uint64 length = 1;
  optional uint64 tailLength = 2;
  optional uint64 numberOfRows = 3;
}

message UserMetadataItem {
  required string name = 1;
  required bytes value = 2;
}

enum CompressionKind {
  NONE = 0;
  ZLIB = 1;
  SNAPPY = 2;
  LZO = 3;
}

message Footer {
  optional uint64 bodyLength = 1;
  optional uint64 stripeOffset = 2;
  repeated StripeInformation stripes = 3;
  repeated Type types = 4;
  repeated UserMetadataItem metadata = 5;
  optional uint64 numberOfRows = 6;
  repeated ColumnStatistics statistics = 7;
}

// Serialized length must be less that 255 bytes
message PostScript {
  optional uint64 footerLength = 1;
  optional CompressionKind compression = 2;
  optional uint64 compressionBlockSize = 3;
}

message StripeSection {
  enum Kind {
    BINARY_DATA = 0;
    BINARY_LENGTH = 1;
    INT_ROW_DATA = 2;
    FLOAT_ROW_DATA = 3;
    DOUBLE_ROW_DATA = 4;
    PRESENT = 5;
    REPETITION = 6;
    DICTIONARY_DATA = 7;
    DICTIONARY_LENGTH = 8;
    DICTIONARY_ROWS = 9;
    DICTIONARY_COUNT = 10;
  }
  required Kind kind = 1;
  optional uint32 column = 2;
}

message ColumnStatistics {
  required uint32 column = 1;
  optional uint64 numberOfValues = 2;
  optional bytes minimum = 3;
  optional bytes maximum = 4;
}

message StripeFooter {
  repeated StripeSection sections = 1;
  repeated ColumnStatistics statistics = 2;
}
